packages/ai-chat/src/browser/session-summary-variable-contribution.ts

```diff
@@ -0,0 +1,115 @@
+        return selection?.id;
+    }
+
+    protected async provideCompletionItems(
+        model: monaco.editor.ITextModel,
+        position: monaco.Position,
+        matchString?: string
+    ): Promise<monaco.languages.CompletionItem[] | undefined> {
+        const context = AIVariableCompletionContext.get(SESSION_SUMMARY_VARIABLE.name, model, position, matchString);
+        if (!context) { return undefined; }
+        const { userInput, range, prefix } = context;
+        return this.getItems().filter(candidate => QuickPickItem.is(candidate) && candidate.label.startsWith(userInput)).map(({ label, id }: QuickPickItem) => ({
+            label,
+            kind: monaco.languages.CompletionItemKind.Class,
+            range,
+            insertText: `${prefix}${id}`,
+            detail: id,
+            filterText: userInput,
+        }));
+    }
+
+    protected getItems(): QuickPickItemOrSeparator[] {
+        return [
+            ...(this.summaries.size ? [{ type: 'separator', label: 'Recent Summaries' }] satisfies QuickPickSeparator[] : []),
+            ...Array.from(this.summaries.entries(), ([id, { label }]) => ({
+                type: 'item',
+                label,
+                id
+            })) satisfies QuickPickItem[],
+            ...(this.summaries.size ? [{ type: 'separator', label: 'Other Sessions' }] satisfies QuickPickSeparator[] : []),
+            ...this.chatService.getSessions()
+                .filter(candidate => !this.summaries.has(candidate.id) && candidate.model.getRequests().length)
+                .map<QuickPickItem>(session => ({ type: 'item', label: session.title || session.id, id: session.id }))
+        ];
+    }
+
+    canResolve(request: AIVariableResolutionRequest, context: AIVariableContext): MaybePromise<number> {
+        return request.variable.id === SESSION_SUMMARY_VARIABLE.id ? 10000 : -5;
+    }
+
+    async resolve(request: AIVariableResolutionRequest, context: AIVariableContext): Promise<ResolvedAIContextVariable | undefined> {
+        if (request.variable.id !== SESSION_SUMMARY_VARIABLE.id || !request.arg) { return; }
+        const existingSession = this.chatService.getSession(request.arg);
+        const existingSummary = this.summaries.get(request.arg);
+        const newSummaryPossibleAndNecessary = !!existingSession && (!existingSummary || existingSummary.length !== existingSession.model.getRequests().length);
+        try {
+            const value = newSummaryPossibleAndNecessary ? await this.summarizeSession(existingSession) : existingSummary?.summary;
+            return value ? { ...request, value, contextValue: value } : undefined;
+        } catch (err) {
+            console.warn('Error retrieving chat session summary for session', request.arg, err);
+            return;
+        }
+    }
+
+    protected async summarizeSession(session: ChatSession): Promise<string> {
+        const summary = await this.summaryAgent.generateChatSessionSummary(session);
```

Comment:
Question 5: We discussed adding a function summarizeSessionToTaskContext (with a better name maybe) to the TaskContextService. As parameters, it would get the chat session and optionally a prompt. The reason for the prompt is that there might be different focusses in the summarization, e.g.:

Summarize what has been done from coder for a follow up
Summarize the plan for Coder from the Architect

Both would probably have slightly different prompts.

Reply:
The TaskContextService accepts arguments for promptId's and agents. At the moment, it's assumed that the prompts will be known to the prompt service, but we could instead accept fully formed messages and not worry whether they came from prompts, and leave that to the caller?

Reply:
I like this a lot for now!